UsageStatsLoader.inc.php

```diff
@@ -416,13 +416,32 @@ protected function getAssoc($assocType, $contextPaths, $page, $op, $args) {
 		$assocId = $assocTypeToReturn = null;
 		switch ($assocType) {
 			case ASSOC_TYPE_SUBMISSION:
+				$publicationId = null;
 				if (!isset($args[0])) break;
+				// If the operation is 'view' and the arguments count > 1
+				// the arguments must be: $submissionId/version/$publicationId.
+				// Else, it is not considered hier, as submission abstract count.
+				if ($op == 'view' && count($args) > 1) {
+					if ($args[1] !== 'version') break;
+					else if (count($args) != 3) break;
+					$publicationId = (int) $args[2];
+				}
 				$submissionId = $args[0];
 				$submissionDao = DAORegistry::getDAO('SubmissionDAO'); /* @var $submissionDao SubmissionDAO */
 				$submission = $submissionDao->getById($submissionId);
 				if ($submission) {
-					$assocId = $submission->getId();
-					$assocTypeToReturn = $assocType;
+					if ($publicationId) {
+						foreach ((array) $submission->getData('publications') as $publication) {
+							if ($publication->getId() === $publicationId) {
+								$assocId = $submission->getId();
+								$assocTypeToReturn = $assocType;
+								break;
+							}
+						}
```

Comment:
What's the case that's being checked here? Are you seeing whether the URL points to a version that doesn't exist? Would that have resulted in an entry usage log at all?

Getting the full submission from the database at line 431 will be expensive. If we can avoid it (ie, by just checking if the submission ID exists instead), that would be better.

Reply:
Yes, you are right! No need to check for that -- having a wrong publication ID would result in 404. Thus, I can remove that check also from here:https://github.com/pkp/ojs/pull/3051/files#diff-fbe83c30f1a2a3e98c218f8398e0c1e7e455de055819ae18014a995024f13e37R106

OK, I'll also only check if the submission ID exists... Maybe then also for other checks? -- similarly we check if context, issue, etc. exist...

Reply:
@NateWr, to check if an objectId (e.g. submissionId) exists I would need to implement that function in each DAO. Is there maybe a better way to do that in the service classes (e.g. getCount() that would accept the own ID)?

Reply:
Maybe leave it as-is for `stable-3_3_0`. But yeah, going forward every entity should have an `exists()` method like the `get()` method for faster checks. I'll think about working this into the Command/Query architecture that we're moving towards in 3.4.

So I'd say for the `main` branch maybe just write a one-off DAO method for this. We can solve the other performance issues for other objects when we look at our broader stats refactor.